src/man/man3/sam_export.3
.\" $Revision: 1.20 $
.ds ]W Sun Microsystems
.\" SAM-QFS_notice_begin
.\"
.\" CDDL HEADER START
.\"
.\" The contents of this file are subject to the terms of the
.\" Common Development and Distribution License (the "License").
.\" You may not use this file except in compliance with the License.
.\"
.\" You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
.\" or https://illumos.org/license/CDDL.
.\" See the License for the specific language governing permissions
.\" and limitations under the License.
.\"
.\" When distributing Covered Code, include this CDDL HEADER in each
.\" file and include the License file at pkg/OPENSOLARIS.LICENSE.
.\" If applicable, add the following below this CDDL HEADER, with the
.\" fields enclosed by brackets "[]" replaced with your own identifying
.\" information: Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" CDDL HEADER END
.\"
.\" Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
.\" Use is subject to license terms.
.\"
.\" SAM-QFS_notice_end
.TH sam_export 3 "07 Jan 2009"
.SH NAME
sam_export \- Exports media from the removable media robotic device
.SH SYNOPSIS
.LP
.BI "cc [ " "flag"
.BI " ... ] " "file"
.BI " ... -L/opt/SUNWsamfs/lib -lsamapi [ " "library" " ... ]"
.LP
.nf
.ft 3
#include "/opt/SUNWsamfs/include/samapi.h"
.ft
.fi
.LP
.BI "int sam_export(ushort_t " "eq_number" ,
.BI "char *" "vsn" ,
.BI "uint_t " "ea" ,
.BI "int " "wait_response" ,
.BI "int " "one_step" );
.SH DESCRIPTION
.PP
.B sam_export(\|)
requests the robotic media changer to place the selected media that belongs
in
.I ea
or is labeled with
.I vsn
into the mail-slot of the media changer.  Either the
.I vsn
or the
.I ea
must be specified.  The
.I vsn
should be set to a NULL pointer if
.I ea
is specified.  The
.I ea
should be set to
.B ROBOT_NO_SLOT
if the
.I vsn
is specified.
.PP
The call will return immediately after issuing the command if zero is
specified for
.I wait_response
value.  Other values for
.I wait_response
will give undefined results.
.PP
For network-controlled media changers such as the GRAU using the GRAU ACT
interface, IBM 3494, or Sony libraries using PSC, this interface only
removes the entry from the catalog.  Physical removal and addition of media
within these media changers is performed by utilities supplied by GRAU, IBM,
and Sony. For STK network-controlled media changers using ACSLS, 
this interface removes the entry from the catalog only if zero is 
specified for 
.I one_step
value. Other values for
.I one_step
will cause the physical removal of media and also remove the media from
the catalog.
.SH "RETURN VALUES"
Upon succesful completion a value of 0 is returned.
Otherwise, a value of \-1 is returned and
\f4errno\fP
is set to indicate the error.
.SH ERRORS
.PP
.B sam_export(\|)
fails if one or more of the following are true:
.TP 20
.SB ER_FIFO_PATH_LENGTH
The path and filename for the FIFO pipe is too long.
.TP
.SB ER_NO_EQUIP_ORDINAL
.I eq_number
is not a defined equipment number in this configuration.
.TP
.SB ER_NO_MASTER_SHM
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master
shared memory segment defined.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS
file systems are mounted.
.TP
.SB ER_NO_MASTER_SHM_ATT
No Sun StorEdge \%SAM-FS or Sun \%SAM-QFS master
shared memory segment found.
Check that the Sun StorEdge \%SAM-FS or Sun \%SAM-QFS
file systems are mounted.
.TP
.SB ER_NO_RESPONSE_FIFO
Unable to create the response FIFO pipe.
.TP
.SB ER_NO_STAT_ROBOT_CATALOG
The robot media changer catalog for equipment number
.I eq_number
cannot be accessed for status.
.TP
.SB ER_NOT_VALID_SLOT_NUMBER
.I ea
is not a valid element address in the robot at equipment number
.I eq_number
.TP
.SB ER_OPERATOR_NOT_PRIV
Operator does not have permission to export removable media.
.TP
.SB ER_ROBOT_CATALOG_MISSING
The robot media changer catalog for equipment number
.I eq_number
is missing and a full audit is required.
.TP
.SB ER_ROBOT_DEVICE_REQUIRED
Equipment number
.I eq_number
is not defined as a robotic device.
.TP
.SB ER_SLOT_NOT_OCCUPIED
.I ea
does not contain any removable media.
.TP
.SB ER_SLOT_OR_VSN_REQUIRED
Either
.I ea
or
.I vsn
must be specified
.TP
.SB ER_UNABLE_TO_MAP_CATALOG
The catalog for the removable media changer at equipment number
.I eq_number
is unable to be mapped into memory.
.TP
.SB ER_VSN_NOT_FOUND_IN_ROBOT
.I vsn
specified cannot be found in the specified robot at equipment number
.I eq_number
.SH FILES
.TP 20
.SB mcf
The configuration file for Sun StorEdge \%SAM-FS and Sun \%SAM-QFS file systems.
.SH SEE ALSO
.BR build_cat (8),
.BR dump_cat (8),
.BR export (8),
.BR import (8),
.BR sam-robotsd (8).
.PP
.BR sam_import (3).
.PP
.BR mcf (5).


src/lib/samapi/sam_export.c
/*
 * sam_export.c - export removable media API
 *
 *	sam_export() causes the media specified to be output from the removable
 *	media device.
 *
 */


/*
 *    SAM-QFS_notice_begin
 *
 * CDDL HEADER START
 *
 * The contents of this file are subject to the terms of the
 * Common Development and Distribution License (the "License").
 * You may not use this file except in compliance with the License.
 *
 * You can obtain a copy of the license at pkg/OPENSOLARIS.LICENSE
 * or https://illumos.org/license/CDDL.
 * See the License for the specific language governing permissions
 * and limitations under the License.
 *
 * When distributing Covered Code, include this CDDL HEADER in each
 * file and include the License file at pkg/OPENSOLARIS.LICENSE.
 * If applicable, add the following below this CDDL HEADER, with the
 * fields enclosed by brackets "[]" replaced with your own identifying
 * information: Portions Copyright [yyyy] [name of copyright owner]
 *
 * CDDL HEADER END
 */
/*
 * Copyright 2009 Sun Microsystems, Inc.  All rights reserved.
 * Use is subject to license terms.
 *
 *    SAM-QFS_notice_end
 */

#pragma ident "$Revision: 1.29 $"

static char *_SrcFile = __FILE__;   /* Using __FILE__ makes duplicate strings */

/* ANSI C headers. */
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* POSIX headers. */
#include <sys/types.h>

/* SAM-FS headers. */
#include "pub/sam_errno.h"
#include "sam/types.h"
#include "sam/param.h"
#include "aml/catalog.h"
#include "aml/catlib.h"
#include "sam/custmsg.h"
#include "aml/device.h"
#include "sam/exit.h"
#include "aml/fifo.h"
#include "sam/lib.h"
#include "aml/proto.h"
#include "aml/samapi.h"
#include "aml/sam_utils.h"

/* Private functions. */
static int ExportCartridge(struct VolId *vid, int wait_response, int OneStep,
		void (*MsgFunc)(int code, char *msg));
static void NoMsgFunc(int code, char *msg);

#ifdef MAIN
int main(int argc, char **argv);
#endif

#ifdef MAIN
int
main(int argc, char **argv)
{
	ushort_t	eq_ord;
	int			slot;
	int			wait_response;
	int			one_step;
	int			error;
	char		*vsn;

	if (argc < 3 || argc > 6) {
		fprintf(stderr, "Usage: %s eq_ord slot "
		    "[ wait_response ] [ one_step ] [ vsn ].\n",
		    argv[0]);
		exit(-70);
	}

	eq_ord		= (ushort_t)atoi(argv[1]);
	slot		= atoi(argv[2]);

	if (argc > 3) {
		wait_response = atoi(argv[3]);
	} else {
		wait_response = 0;
	}

	if (argc > 4) {
		one_step =  atoi(argv[4]);
	} else {
		one_step = 0;
	}

	if (argc > 5) {
		vsn = argv[5];
	} else {
		vsn = (char *)NULL;
	}

	error = sam_export(eq_ord, vsn, slot, wait_response, one_step);

	fprintf(stderr,
	    "Called sam_export to export %d:%d ", (int)eq_ord, slot);

	if (vsn) {
		fprintf(stderr, "vsn %s ", vsn);
	}

	if (wait_response) {
		fprintf(stderr, "wait-response enabled");
	} else {
		fprintf(stderr, "wait-response not enabled");
	}

	if (one_step) {
		fprintf(stderr, " one-step export");
	} else {
		fprintf(stderr, " export-to-historian");
	}

	fprintf(stderr, ".\n");

	fprintf(stderr, "sam_export returned the value %d.  "
	    "errno has the value %d.\n",
	    error, errno);

	if (errno == 0) {
		fprintf(stderr, "The errno %d implies success.\n", (int)0);
	} else {
		char *err_msg;

		err_msg = StrFromErrno(errno, (char *)NULL, 0);

		fprintf(stderr,
		    "The errno %d corresponds to the error message: %s\n",
		    errno, err_msg);
	}

	return (errno);
}
#endif

/*
 * SamExportCartridge - Export cartridge from a library.
 * Returns  0 if success, -1 if failure.
 *
 */
int
SamExportCartridge(
	char *volspec,		/* Volume specification eq:slot or mt.vsn */
	int WaitResponse,	/* Wait for response */
	int OneStep,		/* ACSLS one step export */
	void (*MsgFunc)(int code, char *msg))
{
	struct VolId vid;

	/*
	 * Convert specifier to volume identifier.
	 * Validate specifier.
	 */
	if (StrToVolId(volspec, &vid) != 0) {
		LibError(MsgFunc, EXIT_FAILURE, 18207, volspec);
				/* "Volume specification error %s" */
		return (-1);
	}
	return (ExportCartridge(&vid, WaitResponse, OneStep, MsgFunc));
}


/*
 *	sam_export() - API function to export media
 *
 *	Input parameters --
 *		eq_ord		Equipment ordinal
 *		vsn		Volume serial number to be exported
 *				or(char *) NULL
 *		slot		Slot in robot to be exported
 *		wait_response	Wait for command response if nonzero
 *				-1	Wait forever
 *				> zero	Wait this many seconds for response
 *		OneStep		ACSLS one step export
 *				1 = one step export
 *				0 = export to historian only
 *
 *	Output parameters --
 *		errno		Set to error number if error encountered
 *
 *	Return value --
 *		0		If successful in exporting the media
 *		-1		If error encountered,
 *				'errno' set to error number
 *
 */

int
sam_export(
	ushort_t eq_ord,	/* Equipment ordinal */
	char	*vsn,		/* New Volume Serial Number */
	int 	slot,
	int	wait_response,
int one_step)
{
	struct VolId vid;

	if (vsn == NULL) {
		vid.ViFlags = VI_cart;
		vid.ViEq    = eq_ord;
		vid.ViSlot  = slot;
	} else {
		/*
		 * Convert specifier to volume identifier.
		 * Validate specifier.
		 */
		if (StrToVolId(vsn, &vid) != 0) {
			LibError(NoMsgFunc, EXIT_FAILURE, 18207, vsn);
					/* "Volume specification error %s" */
			return (-1);
		}
	}
	return (ExportCartridge(&vid, wait_response, one_step, NoMsgFunc));
}


/*
 * Message function for sam_export().
 * No messages are processed.
 */
/*ARGSUSED0*/
static void
NoMsgFunc(int code, char *msg)
{
}


/*
 * Common export cartridge.
 */
static int
ExportCartridge(
struct VolId *vid,
int WaitResponse,	/* Wait for response */
int OneStep,		/* ACSLS one step export */
void (*MsgFunc)(int code, char *msg))
{
	struct CatalogHdr *ch;
	struct CatalogEntry ced;
	struct CatalogEntry *ce = &ced;
	dev_ent_t	*dev;
	operator_t	operator;	/* Data on operator */
	sam_cmd_fifo_t	cmd_block;
	char		volspec[STR_FROM_VOLID_BUF_SIZE];
	char		*fifo_path;
				/* Path to FIFO pipe, from shared memory */

	if (CatalogInit("SamApi") == -1) {
		LibError(MsgFunc, EXIT_FAILURE, 18211);
				/* "Catalog initialization failed." */
		return (-1);
	}
	if ((ce = CatalogCheckSlot(vid, &ced)) == NULL) {
		LibError(MsgFunc, EXIT_FAILURE, 18207,
		    StrFromVolId(vid, volspec, sizeof (volspec)));
				/* "Volume specification error %s" */
		return (-1);
	}
	ch = CatalogGetHeader(ce->CeEq);
	if (ch == NULL) {
		LibError(MsgFunc, EXIT_FAILURE, 18207,
		    StrFromVolId(vid, volspec, sizeof (volspec)));
				/* "Volume specification error %s" */
		return (-1);
	}
	if (ch->ChType == CH_manual) {
		LibError(MsgFunc, EXIT_FAILURE, 18207, "Manual");
				/* "Volume specification error %s" */
		return (-1);
	}

	/*
	 * Get device entry and FIFO path from master shared memory segment
	 */
	if (sam_get_dev(ce->CeEq, &dev, &fifo_path, &operator) < 0) {
		LibError(MsgFunc, EXIT_FAILURE, 18213);
		return (-1);
	}

	/*
	 * If operator does not have privilege to move media in robot, return
	 */
	if (!SAM_OPER_SLOT(operator)) {
		errno = ER_OPERATOR_NOT_PRIV;
		LibError(MsgFunc, EXIT_FAILURE, 18213);
		return (-1);
	}

	if (dev->state > DEV_IDLE) {
		if (CatalogExport(vid) == -1) {
			LibError(MsgFunc, EXIT_FAILURE, 18213);
			return (-1);
		}
	}

	/*
	 * Fail if this robot doesn't have a mailbox.
	 * (probably missing some types - Ampex, 3570, etc.)
	 */
	if (dev->equ_type == DT_DLT2700 || dev->equ_type == DT_METD28) {
		LibError(MsgFunc, EXIT_FAILURE, 18213);
		if (!errno) {
			errno = ER_ROBOT_NO_EXPORT_SUPPORT;
		}
		return (-1);
	}

	memset(&cmd_block, 0, sizeof (cmd_block));

	/*
	 * Must define cap to accept exported
	 * volume in the stk params file.
	 */
	if (OneStep) {
		if (dev->dt.rb.capid == ROBOT_NO_SLOT) {
			LibError(MsgFunc, 0, 16077);
			errno = ER_CAPID_NOT_DEFINED;
			return (-1);
		} else {
			cmd_block.flags = CMD_EXPORT_ONESTEP;
		}
	} else {
		cmd_block.flags = 0;
	}

	/*
	 * Format and send commmand to sam-amld
	 */
	cmd_block.magic =  CMD_FIFO_MAGIC;
	cmd_block.cmd =    CMD_FIFO_REMOVE_S;
	cmd_block.eq =	ce->CeEq;
	cmd_block.slot =   ce->CeSlot;

	if (sam_send_cmd(&cmd_block, WaitResponse, fifo_path) < 0) {
		LibError(MsgFunc, EXIT_FAILURE, 16075);
		if (!errno) {
			errno = ER_DEVICE_OP_FAILED;
		}
		return (-1);
	}
	return (0);
}
